File: bin.cpp
#include "bin.h"
bin::bin(int d){
    if(d>=0)
        segno = 0;
    else{
        segno = 1;
        d = -1 * d;
    }

    int quoziente = d;
    bool resto = 0;

    if(!d)
        intera.push_back(0);

    while(quoziente != 0){
        resto = quoziente%2;
        intera.push_back(resto);
        quoziente = quoziente/2;
    }
    std::reverse(intera.begin(), intera.end());
}

double bin::base10() const{
    int i_out = 0;
    int i = 0;
    for(std::vector<bool>::const_reverse_iterator it = intera.rbegin(); it != intera.rend(); ++it){
        i_out = i_out + (*it * (pow(2,i)));
        i++;
    }
    if(segno)
        i_out = -1 * i_out;
    return i_out;
}
std::string bin::base16() const{
    int a = (*this).base10();
    std::string out = "";

    int quoziente = a;
    int resto = 0;

    while(quoziente != 0){
        resto = quoziente%16;
        if(resto < 10)
            out = out + std::to_string(resto);
        else{
            switch (resto)
            {
                case 10:
                    out = out + "A";
                break;

                case 11:
                    out = out + "B";
                break;

                case 12:
                    out = out + "C";
                break;

                case 13:
                    out = out + "D";
                break;

                case 14:
                    out = out + "E";
                break;

                case 15:
                    out = out + "F";
                break;
            }
        }
        quoziente = quoziente/16;
    }
    std::reverse(out.begin(), out.end());
    return out;
}

unsigned int bin::getDimIntera() const { return intera.size(); }

bin operator +(const bin& bin1, const bin& bin2){
    double a = bin1.base10();
    double b = bin2.base10();
    bin out(a+b);
    return out;
}
bin operator -(const bin& bin1, const bin& bin2){
    double a = bin1.base10();
    double b = bin2.base10();
    bin out(a-b);
    return out;
}
bin operator *(const bin& bin1, const bin& bin2){
    double a = bin1.base10();
    double b = bin2.base10();
    bin out(a*b);
    return out;
}
bin operator *(int i, const bin& bin2){
    double a = (double)i;
    double b = bin2.base10();
    bin out(a*b);
    return out;
}
bin operator /(const bin& bin1, const bin& bin2){
    double a = bin1.base10();
    double b = bin2.base10();

    if(b<0)
        throw input_error("bin");

    bin out(a/b);
    return out;
}
bin operator/ (int i, const bin& bin2){
    double a = (double)i;
    double b = bin2.base10();


    if(b==0)
        throw input_error("bin");

    bin out(a/b);
    return out;
}
bool operator< (const bin& bin1, const bin& bin2){
    double a = bin1.base10();
    double b = bin2.base10();

    return a<b;
}
bool operator<= (const bin& bin1, const bin& bin2){
    double a = bin1.base10();
    double b = bin2.base10();

    return a<=b;
}
bool operator> (const bin& bin1, const bin& bin2){
    double a = bin1.base10();
    double b = bin2.base10();

    return a>b;
}
bool operator>= (const bin& bin1, const bin& bin2){
    double a = bin1.base10();
    double b = bin2.base10();

    return a>=b;
}
bool operator== (const bin& bin1, const bin& bin2){
    return (bin1.intera == bin2.intera) /*&& (bin1.frazionaria == bin2.frazionaria)*/ && (bin1.segno == bin2.segno);
}
std::ostream& operator<< (std::ostream& os, const bin& bin1){
    if(bin1.segno)
        os<<"-";
    else
        os<<"+";

    for(unsigned int i=0; i<bin1.intera.size(); i++)
        os<<bin1.intera[i];

    return os;
}

string to_string(const bin& bin1){
    std::string out;

    if(bin1.segno)
        out = "-";
    else
        out = "+";

    for(unsigned int i=0; i<bin1.intera.size(); i++)
        out = out + std::to_string(bin1.intera[i]);

    return out;
}

/* Metodi ereditati dalla classe tipo.h */

bin* bin::somma(tipo* b){
    if(!b)
        throw point_error("bin");
    bin* b1 = dynamic_cast<bin*>(b);
    return new bin((*this) + (*b1));
}
bin* bin::differenza(tipo* b){
    if(!b)
        throw point_error("bin");
    bin* b1 = dynamic_cast<bin*>(b);
    return new bin((*this) - (*b1));
}
bin* bin::moltiplicazione(int b){
    return new bin(b * (*this));    /*è da controllare*/
}
bin* bin::divisione(int b){
    return new bin(b / (*this));    /*è da controllare*/
}
bool bin::uguale(tipo* b) const{
    if(!b)
        throw point_error("bin");
    bin* b1 = dynamic_cast<bin*>(b);
    return (*this)==(*b1);
}
bool bin::min(tipo* b) const{
    if(!b)
        throw point_error("bin");
    bin* b1 = dynamic_cast<bin*>(b);
    return (*this)<(*b1);
}
string bin::to_stringa() const{
    return to_string((*this));
}

File: Java_Kalk/src/dati/Bin.java
/**
 * Classe per rappresentare stringhe binarie. E' presente un ArrayList per
 * rappresentare la parte intera ed uno per la parte frazionaria. Inoltre è
 * un boolean per indicarne il segno
 */
package dati;
import java.util.ArrayList;
import java.lang.Math;
import java.lang.ArithmeticException;
import java.lang.ClassCastException;

public class Bin implements Tipo{
	private ArrayList<Boolean> intera=new ArrayList<Boolean>();
	private ArrayList<Boolean> frazionaria=new ArrayList<Boolean>();
	private Boolean segno;
	public Bin(int d) {
		if(d>=0)
			segno=Boolean.FALSE;
		else {
			segno=Boolean.TRUE;
			d=-d;
		}
		
		int quoziente=d;
		Boolean resto=Boolean.FALSE;
		
		while(quoziente!=0) {
			resto= quoziente%2==0 ? Boolean.FALSE : Boolean.TRUE;
			intera.add(0, resto);
			quoziente=quoziente/2;
		}
		frazionaria.add(Boolean.FALSE);
	}
	public Bin() {
		this(1);
	}
	public Bin(double d) {
		if(d>=0)
			segno=Boolean.FALSE;
		else {
			segno=Boolean.TRUE;
			d=-d;
		}
		
		int intero=(int) d;
		double frazionario=d-intero;
		int quoziente=intero;
		Boolean resto=Boolean.FALSE;
		
		while(quoziente!=0) {
			resto=quoziente%2==0 ? Boolean.FALSE : Boolean.TRUE;
			intera.add(0, resto);
			quoziente=quoziente/2;
		}
		
		if(frazionario==0)
			frazionaria.add(Boolean.FALSE);
		else {
			int i=0;
			while(frazionario!=0 && i<10) {
				frazionario=frazionario*2;
				
				if(frazionario>=1) {
					frazionaria.add(Boolean.TRUE);
					frazionario-=1;
				}
				else
					frazionaria.add(Boolean.FALSE);
				i++;
			}
		}
	}
	//conversione da numero binario a decimale
	public double base10() {
		int i_out=0;
		double f_out=0;
		for(int i=intera.size(); i>0; i--) {
			//l'hashCode ritorna 1231 se intera[i]==true, 1237 se intera[i]==false
			int count=intera.get(i-1).hashCode()-1231==0 ? 1 : 0;
			i_out+=count*Math.pow(2, intera.size()-i);
		}
		
		for(int i=0; i<frazionaria.size(); i++) {
			int count=frazionaria.get(i).hashCode()-1231==0 ? 1 : 0;
			f_out+=count*Math.pow(2, -(i+1));
		}
		
		f_out+=i_out;
		
		if(segno)
			f_out=-f_out;
		
		return f_out;
	}
	//conversione da numero binario a esadecimale
	public String base16() {
		int quoziente=(int)base10();
		String out=new String();
		
		int resto=0;
		
		while(quoziente!=0) {
			resto=quoziente%16;
			if(resto<10)
				out=String.valueOf(resto)+out;
			else {
				switch (resto) {
				case 10:
					out="A"+out;
					break;
					
				case 11:
					out="B"+out;
					break;
					
				case 12:
					out="C"+out;
					break;
				case 13:
					out="D"+out;
					break;
				case 14:
					out="E"+out;
					break;
				case 15:
					out="F"+out;
					break;
				}
			}
			quoziente=quoziente/16;
		}
		
		return out;
	}
	public int getDimIntera() {return intera.size();}
	public int getDimFrazionaria() {return frazionaria.size();}
	public Bin moltiplicazione(Bin bin) {
		double a=base10();
		double b=bin.base10();
		return new Bin(a*b);
	}
	public Bin divisione(Bin bin) throws ArithmeticException{
		double a=base10();
		double b=bin.base10();
		
		if(b<0)
			throw new ArithmeticException("divisione per 0");
		
		return new Bin(a/b);
	}
	
	public Tipo somma(Tipo t) throws ClassCastException{
		if(!(t instanceof Bin))
			throw new ClassCastException("tipi incompatibili per l'operazione");
			
		Bin bin=(Bin)t;
		double a=base10();
		double b=bin.base10();
		return new Bin(a+b);
	}
	public Tipo differenza(Tipo t) throws ClassCastException{
		if(!(t instanceof Bin))
			throw new ClassCastException("tipi incompatibili per l'operazione");
		
		Bin bin=(Bin)t;
		double a=base10();
		double b=bin.base10();
		return new Bin(a-b);
	}
	public Tipo moltiplicazione(int n) {
		return moltiplicazione(new Bin(n));
	}
	public Tipo divisione(int n) {
		return divisione(new Bin(n));
	}
	public boolean equals(Tipo t) throws ClassCastException{
		if(!(t instanceof Bin))
			throw new ClassCastException("tipi incompatibili per l'operazione");
		
		Bin bin=(Bin)t;
		for(int i=0; i<intera.size(); i++)
			if((intera.size()!=bin.intera.size()) || !(intera.get(i).equals(bin.intera.get(i))))
				return false;
		
		for(int i=0; i<frazionaria.size(); i++)
			if((intera.size()!=bin.intera.size()) || !(frazionaria.get(i).equals(bin.frazionaria.get(i))))
				return false;
		return true;
	}
	public boolean min(Tipo t) throws ClassCastException{
		if(!(t instanceof Bin))
			throw new ClassCastException("tipi incompatibili per l'operazione");
		
		Bin bin=(Bin)t;
		double a=base10();
		double b=bin.base10();
		
		return a<=b;
	}
	public String toString() {
		String out=new String();
		
		if(segno)
			out+="-";
		else
			out+="+";
		
		for(int i=0; i<intera.size(); i++) {
			int count=intera.get(i).hashCode()-1231==0 ? 1 : 0;
			out+=String.valueOf(count);
		}
		
		out+=".";
		
		for(int i=0; i<frazionaria.size(); i++) {
			int count=frazionaria.get(i).hashCode()-1231==0 ? 1 : 0;
			out+=String.valueOf(count);
		}
		
		return out;
	}
}
